lib/mfg.c

```diff
@@ -111,34 +111,63 @@ static void RSA_get0_key(const RSA *r, const BIGNUM **n,
 }
 #endif
 
-/**
- * @brief Get secure boot configurations
- * @param[in]  dev	Switchtec device handle
- * @param[out] state	Current secure boot settings
- * @return 0 on success, error code on failure
- */
-int switchtec_security_config_get(struct switchtec_dev *dev,
-				  struct switchtec_security_cfg_state *state)
+static int secure_config_get(struct switchtec_dev *dev,
+			     struct switchtec_security_cfg_state *state,
+			     struct switchtec_security_cfg_otp_region *otp,
+			     bool *otp_valid)
 {
 	int ret;
+	uint8_t subcmd = 0;
 	struct cfg_reply {
 		uint32_t valid;
 		uint32_t rsvd1;
 		uint64_t cfg;
-		uint32_t  public_key_exponent;
-		uint8_t  rsvd2;
-		uint8_t  public_key_num;
-		uint8_t  public_key_ver;
-		uint8_t  rsvd3;
-		uint8_t  public_key[SWITCHTEC_KMSK_NUM][SWITCHTEC_KMSK_LEN];
-		uint8_t  rsvd4[32];
+		uint32_t public_key_exponent;
+		uint8_t rsvd2;
+		uint8_t public_key_num;
+		uint8_t public_key_ver;
+		uint8_t rsvd3;
+		uint8_t public_key[SWITCHTEC_KMSK_NUM][SWITCHTEC_KMSK_LEN];
 	} reply;
 
-	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET, NULL, 0,
+	if (otp_valid)
+		*otp_valid = false;
+
+	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET_EXT,
+				&subcmd, sizeof(subcmd),
 				&reply, sizeof(reply));
-	if (ret)
+	if (ret && ERRNO_MRPC(errno) != ERR_CMD_INVALID)
 		return ret;
 
+	if (!ret) {
+		if (otp) {
+			otp->basic_valid = !!(reply.valid & BIT(5));
+			otp->basic = !!(reply.valid & BIT(6));
+			otp->mixed_ver_valid = !!(reply.valid & BIT(7));
+			otp->mixed_ver = !!(reply.valid & BIT(8));
+			otp->main_fw_ver_valid = !!(reply.valid & BIT(9));
+			otp->main_fw_ver = !!(reply.valid & BIT(10));
+			otp->sec_unlock_ver_valid = !!(reply.valid & BIT(11));
+			otp->sec_unlock_ver = !!(reply.valid & BIT(12));
+			otp->kmsk_valid[0] = !!(reply.valid & BIT(13));
+			otp->kmsk[0] = !!(reply.valid & BIT(14));
+			otp->kmsk_valid[1] = !!(reply.valid & BIT(15));
+			otp->kmsk[1] = !!(reply.valid & BIT(16));
+			otp->kmsk_valid[2] = !!(reply.valid & BIT(17));
+			otp->kmsk[2] = !!(reply.valid & BIT(18));
+			otp->kmsk_valid[3] = !!(reply.valid & BIT(19));
+			otp->kmsk[3] = !!(reply.valid & BIT(20));
+
+			if (otp_valid)
```

Comment:
Do we really need yet another valid flag separate from the struct? Can't the parsing code just look at the existing valid fields? Failing that why not just put another valid field in the struct instead of passing a bool pointer to the function (which is somewhat ugly).

Reply:
The existing valid fields indicates if the OTP regions are valid or not, they can be 0 or 1. The 'otp_valid' is just used to indicates if the current chip returned the OTP info since old firmware doesn't provide it. So the caller cannot guess if otp contains valid info from the individual valid flags in the struct.

Putting another valid field in the otp_region struct seems weird too as I've put another valid flag in the state_ext struct to indicate if member otp contains valid info. I think we can live with it as it's only internal and as a return we make the API looks good.